daemon/daemon.go

```diff
@@ -1123,16 +1123,25 @@ func NewDaemon(dp datapath.Datapath) (*Daemon, *endpointRestoreState, error) {
 	}
 
 	if option.Config.EnableHealthChecking {
-		health4, health6, err := ipam.AllocateNext("")
-		if err != nil {
-			return nil, restoredEndpoints, fmt.Errorf("unable to allocate health IPs: %s,see https://cilium.link/ipam-range-full", err)
+		if option.Config.EnableIPv4 {
+			health4, err := d.ipam.AllocateNextFamily(ipam.IPv4)
+			if err != nil {
+				return nil, restoredEndpoints, fmt.Errorf("unable to allocate health IPs: %s,see https://cilium.link/ipam-range-full", err)
+			}
+
+			d.nodeDiscovery.localNode.IPv4HealthIP = health4
+			log.Debugf("IPv4 health endpoint address: %s", health4)
 		}
 
-		d.nodeDiscovery.localNode.IPv4HealthIP = health4
-		log.Debugf("IPv4 health endpoint address: %s", health4)
+		if option.Config.EnableIPv6 {
+			health6, err := d.ipam.AllocateNextFamily(ipam.IPv6)
+			if err != nil {
+				return nil, restoredEndpoints, fmt.Errorf("unable to allocate health IPs: %s,see https://cilium.link/ipam-range-full", err)
```

Comment:
There's no de-allocation of the potentially allocated `health4` address in this error path. (Though admittedly it's probably less likely you'll run out of IPv6 addresses than IPv4 addresses and actually hit this..)

Reply:
Yes, this is true for all allocations and a lot of other resources. It is currently assumed that if `NewDaemon()` fails, then the agent process is quit. I can fix it up but it doesn't really solve the overall problem.